docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,241 @@
+
+Declared nullability can be retreived from the declaration symbol (e.g. `IFieldSymbol`, `IMethodSymbol`, etc.).
+
+### Flow State
+
+When the nullable feature is enabled, Roslyn will track the flow state of expressions throughout a method, regardless of what the variable was declared as.
+`SemanticModel` can be used to request flow state information for individual expressions or sub-expressions through the `GetTypeInfo` APIs.
+The `TypeInfo` struct will be augmented with `Nullability` and `ConvertedNullability` information, similar to how it currently has `Type` and `ConvertedType`.
+Nullability of symbols at a location can be retrieved through the `GetSymbolInfo` API: if the expression successfully resolved to a single symbol, the `Symbol` property will have nullability calculated.
+`CandidateSymbols` will not have calculated nullabilities, as we cannot run nullable analysis on code that has failed overload resolution.
+
+### `GetSemanticModel(bool skipNullabilityInformation)`
+
+This API proposal changes the way `SemanticModel` calls behave by default.
+Currently, if semantic information is requested for a `SyntaxNode`, we make an effort to bind as little as possible to provide semantic information about that statement, usually a single statement.
+However, if that semantic information will include `Nullability`, we must bind all statements that come before that statement and run nullability analysis up to and including that statement.
+We believe that this change will be ok for most scenarios, as things such as colorization and analyzers are already running in the background near-constantly, binding pretty much everything on screen and in the file anyway.
+There are a few specific scenarios that will be perf-sensitive, and an overload for `GetSemanticModel()` that skips nullability information will be provided for these.
+It will share an initial-binding `BoundNode` cache with the standard semantic model, so it will get the benefits if something else with the same semantic model requests information, as it would have today.
+
+## New APIs
+
+We add the new APIs for retreiving and interacting with nullability:
+
+```C#
+public enum Nullability
+{
+    NotApplicable,
+    NotComputed,
+    Unknown,
+    NotNull,
+    MayBeNull
+}
+
+#region Declared Nullability
+
+// These can potentially be influenced by calculated flow state, but do not change after initial declaration
+
+public interface IDiscardSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IEventSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IFieldSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface ILocalSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IMethodSymbol : ISymbol
+{
+    Nullability ReturnNullability { get; }
+    ImmutableArray<Nullability> TypeArgumentsNullabilities { get; }
+    Nullability ReceiverNullability { get; }
```

Comment:
this feels super unpleasant.  We appear to be making a *dual* API where you need to have an entirely parallel set of nullability info that goes along with any location in our symbolic API where we might reference a type.  This seems onerous and unpleasant, and it is unlike how we did anything else in teh API.

For example, we do not have:
"ReturnIsPointer", "ReturnIsArray", "ReturnIsTuple", "ReturnIsDynamic" or 
`ImmutableArray<bool> TypeArgumentsAreArrays`, or `ImmutableArray<bool> TypeArgumentsAreDynamic` or `ImmutableArray<TupleNameInfo> TypeArgumentTupleNames`, etc. etc.

Instead of using parallel construction, we have always just conveyed different types as actual different types in roslyn.  This violates that construction.

Reply:
This is very similar to how we do both `ref` and custom modifiers today, that's the model we're taking here.

Reply:
I also don't see what scenario this is fixing.  The compiler needs to know about receiver/return nullability for its own tracking but this seems an odd way to expose this info.  Wouldn't the user use a flow analysis api in this case?